### PR TITLE
core: introduce PickSubchannelArgs

### PR DESCRIPTION
**Pull request by lukaszx0**
This will allow adding new arguments passed to pickSubchannel without significant API changes in core LB API in the future.

Addresses #2737

cc @ejona86 @zhangkun83

### CODE REVIEW

File: core/src/main/java/io/grpc/LoadBalancer2.java [hunk 1]
```diff
@@ -317,6 +322,77 @@ public String toString() {
   }
 
   /**
+   * Provides arguments for a {@link SubchannelPicker#pickSubchannel(PickSubchannelArgs)}.
+   */
+  public abstract static class PickSubchannelArgs {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Since LB impls will provide this object, we should make it convenient to create, probably a final class with a builder.

**Comment by lukaszx0** [comment]
@zhangkun83 regarding `methodDescriptor`, I'd argue that it's very useful to have it and there's already a lot of use cases for it, first that immediately comes to mind - ability to have code recording metrics inside picker, based on method's name. I think we should provide it, it'll be very helpful for implementors, myself included. Don't have very strong opinion, can remove it if you're not convinced.

Addressed other comments, PTAL.

File: core/src/main/java/io/grpc/LoadBalancer2.java [hunk 2]
```diff
@@ -170,13 +171,17 @@ public abstract void handleSubchannelState(
...
+
     /**
      * Make a balancing decision for a new RPC.
      *
-     * @param affinity the affinity attributes provided via {@link CallOptions#withAffinity}
-     * @param headers the headers container of the RPC. It can be mutated within this method.
+     * @param args the pick arguments
      */
-    public abstract PickResult pickSubchannel(Attributes affinity, Metadata headers);
+    public abstract PickResult pickSubchannel(PickSubchannelArgs args);
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: AH yeah, good point. Thanks!

File: core/src/main/java/io/grpc/LoadBalancer2.java (hunk 1 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Good call.

File: core/src/main/java/io/grpc/LoadBalancer2.java [hunk 3]
```diff
@@ -317,6 +322,77 @@ public String toString() {
...
+     * Call options.
+     */
+    public CallOptions getCallOptions() {
+      return CallOptions.DEFAULT;
+    }
+
+    /**
+     * Call metadata.
+     */
+    public abstract Metadata getMetadata();
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: 👍

File: core/src/main/java/io/grpc/LoadBalancer2.java (hunk 1 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Oh sorry, I must have gone out of my mind. The args are *consumed* by LB impls, so an abstract class as described in #2737 is most appropriate.

Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: I'm confused. The final immutable class still makes sense to me in this context, no? Do you anticipate passing different types of `PickSubchannelArgs` to `pickSubchannel` internally? If so when and why?

Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: I guess the only change needed would be to either add interface in `LoadBalancer2` or promote `PickSubchannelArgs` from internal to `io.grpc`. Probably latter, since this will let implementors re-use it in tests.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: `PickSubchannelArgs` is public API, thus it should be in `io.grpc` (preferably as inner of `LoadBalancer2`) instead of `io.grpc.internal`.

Builder pattern exposes the construction of `PickSubchannelArgs` as public API, which is not needed by LB impls because they are only consumers of `PickSubchannelArgs`. The builder API has to define required fields (headers) and enforce it, but there isn't an ideal way to do it (there was a [previous discussion](https://github.com/grpc/grpc-java/pull/2667#discussion_r98262481) on a similar issue). It's better to avoid exposing such a builder unless it's necessary.

Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: How would you unit test your custom impl of `Picker` without access to builder? Exposing only abstract class would require you to implement it which effectively will be just copy of what's used internally because this class is just a simple container.

Anyway, unclear at this point if you think we should revert it back to `abstract` or make it `public final` (as it is now) ?

Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: I am thinking about reverting back to `abstract`. The channel would have its own private implementation which will not have a builder, but just a constructor. Unit tests would extend the abstract class to make fake values. Unfortunately Mockito would be discouraged because any unstubbed methods (e.g., new fields) will return null.

Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: That puts a bit of a burden on implementors when writing tests. Not a strong opinion but I slightly favour builder though I understand your point of view and I'm fine reverting that back. Let me know what do you think works better for you.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: I agree with you it's a bit inconvenient for test writers. Because it appears to me both approaches have about the same amount of problems, I would opt for a simpler API, which is the abstract class.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: And then the class would be small enough to fit in `LoadBalancer2`.

Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Reverted back to abstract. PTAL.

**Comment by lukaszx0** [comment]
> You may want to wait until I merge #2743 before making further changes.

It's OK, I've addressed feedback and will rebase after your migration lands. Thanks!

File: core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        && Objects.equal(method, that.method);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(callOptions, headers, method);
+  }
+
+  @Override
+  public String toString() {
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java]: Done.

File: core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    PickSubchannelArgsImpl that = (PickSubchannelArgsImpl) o;
+    return Objects.equal(callOptions, that.callOptions)
+        && Objects.equal(headers, that.headers)
+        && Objects.equal(method, that.method);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(callOptions, headers, method);
+  }
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java]: Good call, though it convinces me a bit that maybe using builder in this case would've been more convenient

File: core/src/main/java/io/grpc/LoadBalancer2.java [hunk 6]
```diff
@@ -317,6 +325,77 @@ public String toString() {
...
 
   /**
+   * Provides arguments for a {@link SubchannelPicker#pickSubchannel(PickSubchannelArgs)}.
+   */
+  public abstract static class PickSubchannelArgs {
+
+    /**
+     * Call affinity.
+     */
+    public Attributes getAffinity() {
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Done.

File: core/src/main/java/io/grpc/LoadBalancer2.java [hunk 7]
```diff
@@ -170,13 +171,20 @@ public abstract void handleSubchannelState(
...
+
     /**
      * Make a balancing decision for a new RPC.
      *
-     * @param affinity the affinity attributes provided via {@link CallOptions#withAffinity}
-     * @param headers the headers container of the RPC. It can be mutated within this method.
+     * @param args the pick arguments
      */
-    public abstract PickResult pickSubchannel(Attributes affinity, Metadata headers);
+    // TODO(lukasz) make it public once deprecated overload will be removed.
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Doh, I meant _abstract_ 🤦‍♂️

File: core/src/main/java/io/grpc/LoadBalancer2.java [hunk 8]
```diff
@@ -170,13 +171,20 @@ public abstract void handleSubchannelState(
    */
   @ThreadSafe
   public abstract static class SubchannelPicker {
+    @Deprecated
+    public PickResult pickSubchannel(Attributes affinity, Metadata headers) {
+      return pickSubchannel(new DeprecatedPickSubchannelArgs(affinity, headers));
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Ah, good point. Removed.

File: core/src/main/java/io/grpc/LoadBalancer2.java [hunk 9]
```diff
@@ -170,13 +171,20 @@ public abstract void handleSubchannelState(
    */
   @ThreadSafe
   public abstract static class SubchannelPicker {
+    @Deprecated
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer2.java]: Done.

File: core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.CallOptions;
+import io.grpc.LoadBalancer2.PickSubchannelArgs;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+// TODO(lukaszx0) make this class pkg private after removing deprecated API in PickSubchannelArgs
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java]: It is already so.

File: core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java [hunk 11]
```diff
@@ -71,6 +72,7 @@
 
   private static final Attributes.Key<String> FOO = Attributes.Key.of("foo");
   private Attributes affinity = Attributes.newBuilder().set(FOO, "bar").build();
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java]: Remove this?

File: core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java [hunk 12]
```diff
@@ -71,6 +72,7 @@
...
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java]: Only if you verify at the end of the test that `EMPTY_ARGS` never had an interaction, it's fine to mock the args without any stubbing.

File: core/src/test/java/io/grpc/util/RoundRobinLoadBalancer2Test.java [hunk 13]
```diff
@@ -87,6 +87,7 @@
   private Map<EquivalentAddressGroup, Subchannel> subchannels = Maps.newLinkedHashMap();
   private static final Attributes.Key<String> MAJOR_KEY = Attributes.Key.of("major-key");
   private Attributes affinity = Attributes.newBuilder().set(MAJOR_KEY, "I got the keys").build();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/util/RoundRobinLoadBalancer2Test.java]: dittos with `PickFirstBalancerFactory2`

File: grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancer2Test.java [hunk 14]
```diff
@@ -157,7 +157,7 @@ public String parse(InputStream stream) {
   private ArgumentCaptor<SubchannelPicker> pickerCaptor;
   private final SerializingExecutor channelExecutor =
       new SerializingExecutor(MoreExecutors.directExecutor());
-  private final Metadata headers = new Metadata();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
Comment (zhangkun83) [inline grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancer2Test.java]: dittos with PickFirstBalancerFactory2

File: core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java (hunk 10 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java]: Ooops, leftover. Thanks!

File: core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java (hunk 11 above)
Comment (lukaszx0) [inline core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java]: It is still used for `handleResolvedAddresses` calls.

File: core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java (hunk 12 above)
Comment (lukaszx0) [inline core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java]: Added `tearDown` with `verifyNoMoreInteractions`

File: core/src/test/java/io/grpc/util/RoundRobinLoadBalancer2Test.java (hunk 13 above)
Comment (lukaszx0) [inline core/src/test/java/io/grpc/util/RoundRobinLoadBalancer2Test.java]: Added `tearDown` with `verifyNoMoreInteractions`

File: grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancer2Test.java (hunk 14 above)
Comment (lukaszx0) [inline grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancer2Test.java]: Added `tearDown` with `verifyNoMoreInteractions`

**Comment by lukaszx0** [comment]
I didn't realize that LBv2 transition will create such bad conflicts so I pretty much just did it over again - it was just easier to do so some copy-pasting from the old version.

@zhangkun83 PTAL and review again as a whole. Thanks!

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 15]
```diff
@@ -318,7 +318,7 @@ void exitIdleMode() {
 
   // Must be run from channelExecutor
   private void cancelIdleTimer() {
-    if (idleModeTimerFuture != null) {
+    if (idleModeTimerFuture != null && idleModeTimer != null) {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Please don't change this line.
 1. This is not related to the subject of this PR.
 2. `idleModeTimer` should be synchronized with `idleModeTimerFuture` in terms of nullness. If one is null while the other is not, it must be a bug and we want to let it throw instead of hiding it.

File: core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.CallOptions;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+class PickSubchannelArgsImpl extends PickSubchannelArgs {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java]: final

File: core/src/main/java/io/grpc/LoadBalancer.java [hunk 17]
```diff
@@ -169,8 +170,75 @@ public abstract void handleSubchannelState(
...
+     * Call method.
+     */
+    public abstract MethodDescriptor<?, ?> getMethodDescriptor();
+
+    /**
+     * Returns true if the given object is also a {@link PickSubchannelArgs} with an equal
+     * getter values.
+     */
+    @Override
+    public boolean equals(Object o) {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer.java]: final

File: core/src/main/java/io/grpc/LoadBalancer.java [hunk 18]
```diff
@@ -169,8 +170,75 @@ public abstract void handleSubchannelState(
...
+      return Objects.equal(getCallOptions(), that.getCallOptions())
+          && Objects.equal(getHeaders(), that.getHeaders())
+          && Objects.equal(getMethodDescriptor(), that.getMethodDescriptor());
+    }
+
+    /**
+     * Returns a hash code for the subchannel args object.
+     */
+    @Override
+    public int hashCode() {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/LoadBalancer.java]: final

File: core/src/test/java/io/grpc/PickFirstLoadBalancerTest.java [hunk 19]
```diff
@@ -71,6 +73,7 @@
 
   private static final Attributes.Key<String> FOO = Attributes.Key.of("foo");
   private Attributes affinity = Attributes.newBuilder().set(FOO, "bar").build();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/PickFirstLoadBalancerTest.java]: We don't want to set an example for users and let them think they can mock `PickSubchannelArgs` whenever we want. We'd better put a comment here says because this LoadBalancer doesn't use any of the arg fields, as verified in `tearDown()`, it's safe to use a mock without any stubbing. `unusedArgs` seems to be a more accurate name. I would also make it non-static, so that if any tests violates this precondition, we will know exactly which they are.

File: core/src/test/java/io/grpc/util/RoundRobinLoadBalancerTest.java [hunk 20]
```diff
@@ -87,6 +88,7 @@
   private Map<EquivalentAddressGroup, Subchannel> subchannels = Maps.newLinkedHashMap();
   private static final Attributes.Key<String> MAJOR_KEY = Attributes.Key.of("major-key");
   private Attributes affinity = Attributes.newBuilder().set(MAJOR_KEY, "I got the keys").build();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/util/RoundRobinLoadBalancerTest.java]: ditto as in PickFirstLoadBalancerTest

File: grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java [hunk 21]
```diff
@@ -157,7 +157,7 @@ public String parse(InputStream stream) {
   private ArgumentCaptor<SubchannelPicker> pickerCaptor;
   private final SerializingExecutor channelExecutor =
       new SerializingExecutor(MoreExecutors.directExecutor());
-  private final Metadata headers = new Metadata();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
Comment (zhangkun83) [inline grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java]: ditto as in PickFirstLoadBalancerTest

File: core/src/main/java/io/grpc/LoadBalancer.java (hunk 17 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/LoadBalancer.java]: Actually, let's not have equals/hashCode. If you're comparing them, Your Doing It Wrong™. That also prevents having the problem later where it is hard to figure out what to consider "equal".

File: core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java (hunk 5 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java]: Ewww... Um, the throwing problem sort of makes me think we shouldn't implement `toString()` in the abstract class either. Granted, we could call a method within try-catch in order to handle an exception, but it is still pretty icky.

The only reason this is an abstract class instead of interface is to allow adding new methods in the future. So in that way, I don't mind requiring toString to be implemented in each extending class.

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 22]
```diff
@@ -155,7 +155,8 @@ public final ClientStream newStream(MethodDescriptor<?, ?> method, Metadata head
       }
       if (picker != null) {
         while (true) {
-          PickResult pickResult = picker.pickSubchannel(callOptions.getAffinity(), headers);
+          PickResult pickResult = picker.pickSubchannel(
+              new PickSubchannelArgsImpl(method, headers, callOptions));
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Tried to do a quick refactor, some tests were failing. Is it ok to address this in a follow up PR?

File: core/src/main/java/io/grpc/LoadBalancer.java (hunk 17 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer.java]: Done

File: core/src/main/java/io/grpc/LoadBalancer.java (hunk 18 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer.java]: Done

File: core/src/main/java/io/grpc/LoadBalancer.java [hunk 23]
```diff
@@ -169,8 +170,75 @@ public abstract void handleSubchannelState(
      *
      * @param affinity the affinity attributes provided via {@link CallOptions#withAffinity}
      * @param headers the headers container of the RPC. It can be mutated within this method.
+     * @deprecated use {@link #pickSubchannel(PickSubchannelArgs)}
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/LoadBalancer.java]: Good call. Added javadocs.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 15 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Damn! I didn't intend to include this, I was experimenting with some tooling and this slipped somehow. Sorry, reverted.

File: core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java (hunk 16 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java]: Done.

File: core/src/test/java/io/grpc/PickFirstLoadBalancerTest.java (hunk 19 above)
Comment (lukaszx0) [inline core/src/test/java/io/grpc/PickFirstLoadBalancerTest.java]: Good points. Renamed it to `mockArgs`, which stick nicely with other fields and the convention we have.

**Comment by lukaszx0** [comment]
Since comments on diff collapsed, moving the unaddressed one from @ejona86 up for discussion here:

> Actually, let's not have equals/hashCode. If you're comparing them, Your Doing It Wrong™. That also prevents having the problem later where it is hard to figure out what to consider "equal".

I agree on the principle, but turns our our tests are relying on this (as will probably the ones implemented by users).

@ejona86 @zhangkun83 thoughts?

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java (hunk 22 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Hmm, I didn't expect this to be more than a trivial change. What tests broke?

**Comment by ejona86** [comment]
> but turns our our tests are relying on this (as will probably the ones implemented by users)

Then the concrete implementation of PickSubchannelArgs used in the tests could implement the methods. The concrete implementation can easily avoid exceptions and API oddity.

**Comment by lukaszx0** [comment]
> Then the concrete implementation of PickSubchannelArgs used in the tests could implement the methods. The concrete implementation can easily avoid exceptions and API oddity.

Then it's fine to have them in `PickSubchannelArgsImpl` too, right?

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java (hunk 22 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: It's possible that it is and I've made some stupid mistake, the `DelayedClientTransportTest` and one other tests were failing. I'll look into this again later today.

**Comment by lukaszx0** [comment]
@ejona86 @zhangkun83 PTAL

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 24]
```diff
@@ -347,26 +348,25 @@ public LogId getLogId() {
...
-      this.method = method;
-      this.headers = headers;
-      this.callOptions = callOptions;
-      this.context = Context.current();
+    private PendingStream(PickSubchannelArgs args, StatsTraceContext statsTraceCtx) {
+      this.args = args;
       this.statsTraceCtx = statsTraceCtx;
     }
 
+    public PickSubchannelArgs getArgs() {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: This method is not necessary, because `args` is only used from the same file.

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 25]
```diff
@@ -290,24 +291,24 @@ final void reprocess(SubchannelPicker picker) {
...
-            @Override
-            public void run() {
-              stream.createRealStream(transport);
-            }
-          });
+          @Override
+          public void run() {
+            stream.createRealStream(transport);
+          }
+        });
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Please don't do any no-op reformatting.

File: core/src/test/java/io/grpc/internal/DelayedClientTransportTest.java [hunk 26]
```diff
@@ -479,89 +501,92 @@ public void reprocess_newStreamRacesWithReprocess() throws Exception {
...
-          // Will call pickSubchannel and wait on barrier
-          delayedTransport.newStream(method, headers, callOptions, statsTraceCtx);
-        }
-      };
+      @Override
+      public void run() {
+        // Will call pickSubchannel and wait on barrier
+        delayedTransport.newStream(method, headers, callOptions, statsTraceCtx);
+      }
+    };
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/internal/DelayedClientTransportTest.java]: ditto -- please refrain from any no-op reformatting.

File: core/src/test/java/io/grpc/internal/DelayedClientTransportTest.java [hunk 27]
```diff
@@ -479,89 +501,92 @@ public void reprocess_newStreamRacesWithReprocess() throws Exception {
...
-          // Will call pickSubchannel and wait on barrier
-          delayedTransport.newStream(method, headers2, callOptions, statsTraceCtx);
-        }
-      };
+      @Override
+      public void run() {
+        // Will call pickSubchannel and wait on barrier
+        delayedTransport.newStream(method, headers2, callOptions, statsTraceCtx);
+      }
+    };
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/internal/DelayedClientTransportTest.java]: ditto

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java (hunk 24 above)
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Right, fixed.

**Comment by lukaszx0** [comment]
@zhangkun83 reverted formatting that you pointed out and found it in few more places - sorry IJ did that to me ;) Should be good now, PTAL.

**Comment by ejona86** [comment]
I did some spot-checking, and this looks fine to me.

> Then it's fine to have them in `PickSubchannelArgsImpl` too, right?

Ideally, I'd say no, since applications could depend on the behavior. I think @zhangkun83 and I agree on that. However, some of the tests look very painful to fix if equals was dropped. So while disappointed, I'm okay with having equals/hashcode on the primary Impl because doing otherwise would be too onerous.

**Comment by lukaszx0** [comment]
@ejona86 thanks, sounds good! @zhangkun83 are we good to go with this?

**Comment by zhangkun83** [comment]
Please resolve the conflict.

**Comment by lukaszx0** [comment]
Oops, didn't notice there was a conflict. Fixed!
